nomos-services/membership/src/lib.rs

```diff
@@ -124,64 +123,100 @@ where
-                            for (service_type, snapshot) in snapshot {
-                                if let Some(tx) = self.subscribe_txs.get(&service_type) {
-                                    if tx.send(snapshot).is_err() {
-                                        tracing::error!("Error sending membership update");
-                                    }
-                                }
-                            }
-                        }
-                        Err(e) => {
-                            tracing::error!("Failed to update backend: {:?}", e);
-                        }
-                     }
+                     self.handle_sdp_update(sdp_msg).await;
                 },
             }
         }
         Ok(())
     }
 }
 
+impl<B, S, RuntimeServiceId> MembershipService<B, S, RuntimeServiceId>
+where
+    B: MembershipBackend,
+    S: SdpAdapter,
+    B::Settings: Clone,
+{
+    async fn handle_message(&mut self, msg: MembershipMessage) {
+        match msg {
+            MembershipMessage::GetSnapshotAt {
+                reply_channel,
+                block_number,
+                service_type,
+            } => {
+                let result = self
+                    .backend
+                    .get_providers_at(service_type, block_number)
+                    .await;
+                if let Err(e) = reply_channel.send(result) {
+                    tracing::error!("Failed to send response: {:?}", e);
+                }
+            }
+            MembershipMessage::Subscribe {
+                service_type,
+                result_sender,
+            } => {
+                let tx = self
+                    .subscribe_channels
+                    .entry(service_type)
+                    .or_insert_with(|| {
+                        let (tx, _) = broadcast::channel(BROADCAST_CHANNEL_SIZE);
+                        tx
+                    });
+
+                let stream = make_pin_broadcast_stream(tx.subscribe());
+                let providers = self.backend.get_latest_providers(service_type).await;
+
+                if let Ok(providers) = providers {
+                    if tx.send(providers).is_err() {
+                        tracing::error!(
+                            "Error sending initial membership snapshot for service type: {:?}",
+                            service_type
+                        );
+                    } else if result_sender.send(stream).is_err() {
+                        tracing::error!(
+                            "Error sending finalized updates receiver for service type: {:?}",
+                            service_type
+                        );
+                    }
+                } else {
+                    tracing::error!(
+                        "Failed to get latest providers for service type: {:?}",
+                        service_type
+                    );
```

Comment:
Should we sent an empty stream of providers back from the api? Otherwise we would be blocking the consuming service.

Reply:
Hmmm, make sense.. Actually on a second thought, I will return an error, good catch

Reply:
Fixed.